File: content/LLD/Reference-Code/11. Singleton Pattern/good_example.py
class Logger:
    # Class Variable
    __instance = None

    def __new__(cls, file_name: str):
        if cls.__instance == None:
            cls.__instance = super().__new__(cls)
            cls.__instance.file_name = file_name
            cls.__instance.log_count = 0
            return cls.__instance
        else:
            return cls.__instance

    def log(self, msg):
        print(f"Logging {msg} in {self.file_name}")
        self.log_count += 1

    def get_log_count(self) -> int:
        return self.log_count


log1 = Logger("app.log")
log1.log("Hey")

log2 = Logger("app.log")
log2.log("Bye")

log3 = Logger("app.log")
log3.log("Good")

print(log1.get_log_count())
print(log2.get_log_count())
print(log3.get_log_count())

File: content/LLD/Reference-Code/11. Singleton Pattern/GoodExample.java
public class GoodExample {
    public static class Logger {
        private static Logger instance;
        private final String fileName;
        private int logCount;

        private Logger(String fileName) {
            this.fileName = fileName;
            this.logCount = 0;
        }

        public static synchronized Logger getInstance(String fileName) {
            if (instance == null) {
                instance = new Logger(fileName);
            }
            return instance;
        }

        public void log(String msg) {
            System.out.println("Logging " + msg + " in " + this.fileName);
            this.logCount++;
        }

        public int getLogCount() {
            return this.logCount;
        }
    }

    public static void main(String[] args) {
        Logger log1 = Logger.getInstance("app.log");
        log1.log("Hey");

        Logger log2 = Logger.getInstance("app.log");
        log2.log("Bye");

        Logger log3 = Logger.getInstance("app.log");
        log3.log("Good");

        System.out.println(log1.getLogCount());
        System.out.println(log2.getLogCount());
        System.out.println(log3.getLogCount());
    }
}
